lib/src/requests_signature_interceptor.dart

```diff
@@ -0,0 +1,146 @@
+  ///            Optionally, 'hmacSha512' can be provided for custom HMAC algorithm instance.
+  factory HmacInterceptor.fromConfig(Map<String, dynamic> config) {
+    final applicationId = config['applicationId'];
+    final applicationSecret = config['applicationSecret'];
+    final hmacSha512 =
+        config['hmacSha512'] != null ? Hmac(config['hmacSha512']) : null;
+    return HmacInterceptor(
+        applicationId: applicationId,
+        applicationSecret: applicationSecret,
+        hmacSha512: hmacSha512);
+  }
+
+  @override
+  void onRequest(
+      RequestOptions options, RequestInterceptorHandler handler) async {
+    try {
+      final time = _epochTime();
+      final nonce = _newGuid();
+      final method = options.method;
+      final url = options.uri.toString();
+
+      final cacheKey = '$_applicationId:$nonce:$time:$method:$url';
+      final token = _tokenCache[cacheKey] ??
+          await _calculateToken(
+            _applicationId,
+            nonce,
+            time,
+            method,
+            url,
+            _applicationSecret,
+          );
+
+      _tokenCache[cacheKey] = token;
+
+      final hmac = '$_applicationId:$token:$nonce:$time';
+      options.headers.addAll({
+        _headerHmac: hmac,
+        _headerOperationId: _newGuid(),
+      });
+      return super.onRequest(options, handler);
+    } catch (e, stackTrace) {
+      // Handle the exception
+      throw HmacRequestException(
+        message: 'Failed to process request: ${e.toString()}',
+        stackTrace: stackTrace,
+      );
+    }
+  }
+
+  /// Generates a new GUID.
+  ///
+  /// Returns:
+  ///   A new GUID string.
+  String _newGuid() {
+    const String uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
+    return uuid.replaceAllMapped(RegExp('[xy]'), (match) {
+      final int rand = Random.secure().nextInt(16);
+      final int index = match.group(0) == 'x' ? rand : (rand & 0x3 | 0x8);
+      return index.toRadixString(16);
+    });
+  }
+
+  /// Gets the current epoch time in seconds.
+  ///
+  /// Returns:
+  ///   Current epoch time in seconds.
+  String _epochTime() {
+    final int t = DateTime.now().millisecondsSinceEpoch;
+    final String o = t.toString();
+    return o.substring(0, 10);
+  }
+
+  /// Calculates the HMAC token for the request.
+  ///
+  /// [applicationId]: Application ID for HMAC authentication.
+  /// [nonce]: Nonce value for the request.
+  /// [time]: Epoch time of the request.
+  /// [method]: HTTP method of the request.
+  /// [url]: URL of the request.
+  /// [applicationSecret]: Application secret for HMAC authentication.
+  ///
+  /// Returns:
+  ///   HMAC token.
+  Future<String> _calculateToken(
+    String applicationId,
+    String nonce,
+    String time,
+    String method,
+    String url,
+    String applicationSecret,
+  ) async {
+    final uri = Uri.parse(url);
+    final payloadBuffer = StringBuffer()
```

Comment:
I guess this is more efficient than a string?

Reply:
in a way it is yes

Reply:
Cool! Learned a new thing!